File: BernSCM_Procedural/Python/bernSCM_readforcing.py
"""
Brice Petit 000439675
Master Thesis
Transformation of BernSCM in procedural
Python version
"""
import os

import numpy as np

import Globals


def getPath(folder_name):
    """
    Get the complete path and go to the folder corresponding to the folder name.

    :param folder_name: The name of the folder where we want to go.
    :type folder_name:  Str.

    :return:            Return the path.
    :rtype:             Str
    """
    path = os.path.abspath(os.getcwd())[:-6] + folder_name
    if os.name == "nt":
        separator = "\\"
    else:
        separator = "/"
    return path + separator


def readForcing():
    """
    Read forcing file.
    """
    with open(getPath("forcing") + "forcing_" + Globals.scenario + ".dat", "r") as file:
        data = file.readlines()
        for i in range(len(data)):
            splited_tmp = np.array(data[i].split())
            if splited_tmp[0][0] != "#":
                tmp_res = np.array([], dtype=np.longdouble)
                for j in range(Globals.NFORC):
                    if abs(float(splited_tmp[j]) - Globals.NA) < 1e-3:
                        tmp_res = np.append(tmp_res, Globals.NA)
                    elif j == Globals.JACO2:
                        tmp_res = np.append(tmp_res, float(splited_tmp[j]) * Globals.PPMTOGT)
                    else:
                        tmp_res = np.append(tmp_res, float(splited_tmp[j]))
                Globals.forcing = np.append(Globals.forcing, [tmp_res], axis=0)

File: BernSCM_Procedural/Cpp/src/bernSCM_readforcing.cpp
//
// Created by Brice Petit on 25-02-22.
//

#include "bernSCM_readforcing.hpp"

/**
 * Get the complete path and go to the folder corresponding to the folder name.
 *
 * @param folder_name   The name of the folder where we want to go.
 * @return              Return the path.
 */
string getPath(const string &folder_name) {
    string separator, path;
    path = fs::current_path().string();
    for (int i = 0; i < 3; i++) {
        path.pop_back();
    }
    path += folder_name;
#ifdef _WIN32
    separator = "\\";
#else
    separator = "/";
#endif
    return path + separator;
}

/**
 * Read forcing file.
 */
void readForcing() {
    string line, tmp;
    double frecord[NFORC];
    int i = 0, j, k, l;
    ifstream my_file;
    my_file.open(getPath("forcing") + "forcing_" + scenario + ".dat");
    // We count the number of lines in the file.
    while (getline(my_file, line)) {
        if (line[0] != '#' && line[1] != '#') {
            FORCING_ROWS++;
        }
    }
    my_file.clear();
    my_file.seekg(0);
    // Define the number of rows needed.
    forcing = new double *[FORCING_ROWS];
    // We create our forcing matrix.
    while (getline(my_file, line)) {
        if (line[0] != '#' && line[1] != '#') {
            l = 0;
            // We search each value in the line that we read.
            for (k = 0; k <= line.size(); k++) {
                if (line[k] != ' ' && k < line.size()) {
                    tmp += line[k];
                } else if ((!tmp.empty() && line[k] == ' ') || (k == line.size())) {
                    frecord[l] = stod(tmp);
                    l++;
                    tmp = "";
                }
            }
            forcing[i] = new double[NFORC];
            for (j = 0; j < NFORC; j++) {
                if (abs(frecord[j] - NA) < 1e-3) {
                    forcing[i][j] = NA;
                } else if (j == JACO2) {
                    forcing[i][j] = frecord[j] * PPMTOGT;
                } else {
                    forcing[i][j] = frecord[j];
                }
            }
            i++;
        }
    }
    my_file.close();
}
